Fix for RoslynPad crash found in targeted preview testing

trylek opened the pull request:
I found out that the crash was caused by an incompletely handled
corner case in generic lookup - for method generic lookups, we
were always requesting instantiation stubs even when not appropriate.
I have updated the condition for emitting an instantiation stub,
please review it and let me know if you think the check is incomplete.

Thanks

Tomas

Fixes: https://github.com/dotnet/runtime/issues/50472

trylek committed:
Fix for RoslynPad crash found in targeted preview testing

I found out that the crash was caused by an incompletely handled
corner case in generic lookup - for method generic lookups, we
were always requesting instantiation stubs even when not appropriate.
I have updated the condition for emitting an instantiation stub,
please review it and let me know if you think the check is incomplete.

Thanks

Tomas

jkotas commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/GenericLookupSignature.cs:
```diff
@@ -107,12 +107,15 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                 Debug.Assert(_methodArgument.Unboxing == false);
 
+                bool useInstantiatingStub = (_methodArgument.Method.IsSharedByGenericInstantiations &&
+                    (_methodArgument.Method.Signature.IsStatic || _methodArgument.Method.HasInstantiation));
```
Does this check also need to include value types?

Or should this condition be just `RequiresInstArg`?

trylek replied:
I have reverted this part of the change which doesn't work anyway as the method in question is "runtime determined", not "canonical". According to David Wrighton's analysis it turned out the problem had a different cause, namely imprecise comparison of generic context keys that could in very complex cases end up replacing a generic lookup with a different one.

davidwrighton commented:
Please add a testcase for this to crossgen2smoke.

trylek committed:
Fix GenericLookupKey comparison

In our offline discussion David pointed out that I misunderstood
the cause of the failure; the problem was not due to the extra
instantiation stub but due to the fact that incomplete comparison
of GenericLookupKey matched two generic lookups which in reality
involved two different constrained types.

While fixing this I added a new variant to
RuntimeDeterminedTypeHelpers for comparing MethodWithToken and
I fixed null handling that was inconsistent among the individual
type variants of the Equals and GetHashCode methods.

Thanks

Tomas

davidwrighton reviewed (approved):
Now this looks right. I would still like to see a repro case, but I don't want work on that to delay getting this fix into place for the next preview.

trylek commented:
Thanks David for your review. In accordance with your feedback I have merged the change in to give it at least some bake time before the P4 fork. I'll create the regression test as the next step.
